nixos/modules/services/network-filesystems/ipfs.nix

```diff
@@ -9,7 +9,10 @@ let
 
   ipfsFlags = ''${if cfg.autoMigrate then "--migrate" else ""} ${if cfg.enableGC then "--enable-gc" else ""} ${toString cfg.extraFlags}'';
 
-  pathEnv = { IPFS_PATH = cfg.dataDir; };
+  # Before Version 17.09, ipfs would always use "/var/lib/ipfs/.ipfs" as it's dataDir
+  defaultDataDir = if versionAtLeast config.system.stateVersion "17.09" then
+    "/var/lib/ipfs" else
+    "/var/lib/ipfs/.ipfs";
 
```

Comment:
Shouldn't `stateVersion` generally stay the same? If the user decides to change it he'll have to look through what breaks themselves, because I think NixOS doesn't need to provide any stability guarantees with stateful data beyond this option. Also it's just the default, if the user set's it explicitly later they'll have to move it themselves. I think that's really the intended way of using `stateVersion`, citing from the description of the option:

> Every once in a while, a new NixOS release may change configuration defaults in a way incompatible with stateful data. [...] To prevent such breakage, you can set the value of this option to the NixOS release with which you want to be compatible. [...]